### PR TITLE
Fix some name completion bugs

### PR DESCRIPTION
**Pull request by rchande**
Fixes:
https://github.com/dotnet/roslyn/issues/19409
https://github.com/dotnet/roslyn/issues/20016
https://github.com/dotnet/roslyn/issues/20273

Tag @dotnet/roslyn-ide for review

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/DeclarationNameCompletionProviderTests.cs [hunk 1]
```diff
@@ -453,5 +453,96 @@ public class MyClass
...
+            var markup = @"
+class Test
+{
+    void Do(out Test foo)
+    {
+        Do(out var $$
+    }
+}
+";
+            await VerifyItemExistsAsync(markup, "test");
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/DeclarationNameCompletionProviderTests.cs]: nummy!

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/DeclarationNameCompletionProviderTests.cs [hunk 2]
```diff
@@ -453,5 +453,96 @@ public class MyClass
...
+            var markup = @"
+class Test
+{
+    void Do(out Test foo)
+    {
+        Do(out Test $$
+    }
+}
+";
+            await VerifyItemExistsAsync(markup, "test");
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/DeclarationNameCompletionProviderTests.cs]: Can you add a test where it is ```Do<T>(out T foo)``` and the user writes "Do(out Whatever $$".  I would like both "whatever" offered as well as "foo"

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs [hunk 3]
```diff
@@ -59,8 +65,42 @@ internal struct NameDeclarationInfo
...
+                ITypeInferenceService typeInferenceService, CancellationToken cancellationToken, out NameDeclarationInfo result)
+            {
+                if (!token.IsKind(SyntaxKind.IdentifierToken) || !(token.Parent.IsKind(SyntaxKind.IdentifierName)))
+                {
+                    result = default;
+                    return false;
+                }
+
+                var argument = token.Parent.Parent as ArgumentSyntax // Foo(out var $$
+                    ?? token.Parent.Parent.Parent as ArgumentSyntax; // Foo(out var a$$
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs]: not sure i like this.  We have an existing system to say "move to the left one token if i'm on an actual name".  i'd prefer we stick witth htat pattern.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs [hunk 4]
```diff
@@ -59,8 +65,42 @@ internal struct NameDeclarationInfo
...
+
+                var type = typeInferenceService.InferType(semanticModel, argument.SpanStart, objectAsDefault: false, cancellationToken: cancellationToken);
+                if (type != null)
+                {
+                    result = new NameDeclarationInfo(
+                        ImmutableArray.Create(SymbolKind.Local),
+                        Accessibility.NotApplicable,
+                        new DeclarationModifiers(),
+                        type,
+                        null);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs]: named param for null.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.NameGenerator.cs [hunk 5]
```diff
@@ -70,13 +85,18 @@ private static string TryRemoveInterfacePrefix(ITypeSymbol type)
...
                     {
                         return name.Substring(1);
                     }
                 }
                 return type.CreateParameterName();
             }
         }
+
+        private static bool CanRemoveInterfacePrefix(string name)
+        {
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.NameGenerator.cs]: nit: =>

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs (hunk 3 above)
Comment (rchande) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs]: It's actually the same token either way (the token after `out`, eg `var`). However, once you start typing the name (so like `out var foo`), we start parsing that as a DeclarationExpressionSyntax. Without the name, `var` is a direct child of the ArgumentSyntax.

**Comment by rchande** [comment]
retest ubuntu_16_debug_prtest

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/DeclarationNameCompletionProviderTests.cs (hunk 2 above)
Comment (rchande) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/DeclarationNameCompletionProviderTests.cs]: Done.

**Pull request merged**